staging/src/slime.io/slime/modules/lazyload/controllers/servicefence_controller.go

```diff
@@ -62,59 +59,67 @@ type ServicefenceReconciler struct {
-	if err != nil {
-		log.Errorf("%v", err)
-		return nil
+func ReconcilerWithEnv(env bootstrap.Environment) ReconcilerOpts {
+	return func(sr *ServicefenceReconciler) {
+		sr.env = env
+		sr.defaultAddNamespaces = append(sr.defaultAddNamespaces, env.Config.Global.IstioNamespace, env.Config.Global.SlimeNamespace)
 	}
+}
 
-	r := &ServicefenceReconciler{
-		Client:               mgr.GetClient(),
-		Scheme:               mgr.GetScheme(),
-		env:                  env,
-		interestMeta:         map[string]bool{},
-		interestMetaCopy:     map[string]bool{},
-		watcherMetricChan:    pc.WatcherProducerConfig.MetricChan,
-		tickerMetricChan:     pc.TickerProducerConfig.MetricChan,
-		staleNamespaces:      map[string]bool{},
-		enabledNamespaces:    map[string]bool{},
-		defaultAddNamespaces: []string{env.Config.Global.IstioNamespace, env.Config.Global.SlimeNamespace},
-		doAliasRules:         newDomainAliasRules(cfg.DomainAliases),
-		cfg:                  cfg,
-	}
-
-	// start service related cache
-	r.startSvcCache()
-
-	// reconciler defines producer metric handler
-	pc.WatcherProducerConfig.NeedUpdateMetricHandler = r.handleWatcherEvent
-	pc.TickerProducerConfig.NeedUpdateMetricHandler = r.handleTickerEvent
-
-	// start producer
-	metric.NewProducer(pc)
-	log.Infof("producers starts")
-
-	if env.Config.Metric != nil || env.Config.Global.Misc["metricSourceType"] == MetricSourceTypeAccesslog {
-		go r.WatchMetric()
-	} else {
-		log.Warningf("watching metric is not running")
+func ReconcilerWithCfg(cfg *lazyloadv1alpha1.Fence) ReconcilerOpts {
+	return func(sr *ServicefenceReconciler) {
+		sr.cfg = cfg
+		sr.doAliasRules = newDomainAliasRules(cfg.DomainAliases)
+	}
+}
+
+func ReconcilerWithProducerConfig(pc *metric.ProducerConfig) ReconcilerOpts {
+	return func(sr *ServicefenceReconciler) {
+		sr.watcherMetricChan = pc.WatcherProducerConfig.MetricChan
+		sr.tickerMetricChan = pc.TickerProducerConfig.MetricChan
+		// reconciler defines producer metric handler
+		pc.WatcherProducerConfig.NeedUpdateMetricHandler = sr.handleWatcherEvent
+		pc.TickerProducerConfig.NeedUpdateMetricHandler = sr.handleTickerEvent
 	}
+}
 
+// NewReconciler returns a new reconcile.Reconciler
+func NewReconciler(opts ...ReconcilerOpts) *ServicefenceReconciler {
+	r := &ServicefenceReconciler{
+		interestMeta:      map[string]bool{},
+		interestMetaCopy:  map[string]bool{},
+		staleNamespaces:   map[string]bool{},
+		enabledNamespaces: map[string]bool{},
+		nsSvcCache:        &NsSvcCache{Data: map[string]map[string]struct{}{}},
+		labelSvcCache:     &LabelSvcCache{Data: map[LabelItem]map[string]struct{}{}},
+		portProtocolCache: &PortProtocolCache{Data: map[int32]map[Protocol]uint{}},
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
 	return r
 }
 
+func (r *ServicefenceReconciler) Clear() {
+	r.reconcileLock.Lock()
+	defer r.reconcileLock.Unlock()
+	// reset cache
+	r.interestMeta = map[string]bool{}
+	r.interestMetaCopy = map[string]bool{}
+	r.staleNamespaces = map[string]bool{}
```

Comment:
these two cache seems to do nothing with the state transition

Reply:
for now clear do nothing